### DISCUSSION TITLE
Why do my text elements keep disappearing from the output every time a new one is added?

### DISCUSSION THREAD

**Discussion by mscofield0** [ArthurSonzogni/FTXUI#275] [Q&A]
This is roughly what the `add_message()` function looks like:
```cpp
std::string msg;
main_component_->message_component_->Add(ftxui::Renderer([msg] { return ftxui::text(msg); }));
```
And this is what the render function looks like:
```
auto App::Component::Render() -> ftxui::Element {
	using namespace ftxui;

	return vbox({
	  window(text("Logs"), log_component_->Render()) | flex,
	  window(text("Chat"), message_component_->Render()) | flex,
	  window(text("Input"), input_->Render()),
	});
}
```
With the component class being like this:
```cpp
class Component : public ftxui::ComponentBase
{
	friend class App;
public:
	auto Render() -> ftxui::Element override;
	auto OnEvent(ftxui::Event evt) -> bool override;

private:
	ftxui::Component log_component_ = ftxui::Container::Vertical({});
	ftxui::Component message_component_ = ftxui::Container::Vertical({});
	std::string input_buf_;
	ftxui::Component input_ = ftxui::Input(&input_buf_, "Input here:", {});
};
```

**Comment by ArthurSonzogni** [comment]
How is the constructor written?

**Reply by mscofield0** [reply]
```cpp
App::Component::Component() {
	this->Add(log_component_);
	this->Add(message_component_);
	this->Add(input_);
}
```

There is also this where I need to detach and reattach the `Input` component because I want different "on send" functionality for my server and for my client. 
```cpp
auto App::attach(std::function<void()>&& on_send) -> void {
	ftxui::InputOption opt;
	opt.on_enter = std::move(on_send);

	main_component_->input_->Detach();
	main_component_->input_ = ftxui::Input(&main_component_->input_buf_, "Input here:", opt);
	main_component_->Add(main_component_->input_);
}
```

**Reply by ArthurSonzogni** [reply]
This doesn't look bad in any ways :-)

Maybe you could check the number of children is increased correctly?
```cpp
auto App::Component::Render() -> ftxui::Element {
	using namespace ftxui;

	return vbox({
          text("Child count = " + std::to_string(ChildCount())),
	  window(text("Logs"), log_component_->Render()) | flex,
	  window(text("Chat"), message_component_->Render()) | flex,
	  window(text("Input"), input_->Render()),
	});
}
```

**Reply by mscofield0** [reply]
The main component stays on 3 children, but the `message_component_` increases when I press enter.

**Reply by mscofield0** [reply]
Also do you know how I could clear the input component? I tried to clear the `std::string` passed to it, but it doesn't work. I did redraw after it, so it's not that.

**Reply by ArthurSonzogni** [reply]
Yes, sorry. I meant checking the message_component_ one.
So, it increase correctly. Could you describe the problem you are seeing exactly?

```
Also do you know how I could clear the input component? I tried to clear the std::string passed to it, but it doesn't work. I did redraw after it, so it's not that.
```
I am expecting clearing `input_buf_` to clear what is drawn.

**Reply by mscofield0** [reply]
![image](https://user-images.githubusercontent.com/69715407/145107254-d3f95cbd-8fd9-44e0-85a6-2b143465dcdd.png)

After entering a new message:
![image](https://user-images.githubusercontent.com/69715407/145107305-1311b250-f1d2-4df2-a57b-b2f1727311f4.png)

After pressing a couple of keys, but not enter:
![image](https://user-images.githubusercontent.com/69715407/145107371-0cc521ac-1b5d-4ade-b49d-2912bd72e5fe.png)

**Reply by mscofield0** [reply]
> I am expecting clearing input_buf_ to clear what is drawn.

I have no idea why it doesn't work then...

**Reply by mscofield0** [reply]
This is what I attach to the `on_enter` of `Input`:
```cpp
UI::App::instance().attach([] {
	UI::App::instance().add_message(UI::App::instance().input_buffer());
	UI::App::instance().input_buffer().clear();
	UI::App::instance().redraw();
});
```

**Reply by ArthurSonzogni** [reply]
I have no idea.
You can publish your repository for me to take a look if you want.

(I am going to sleep, see me reply tomorrow)

**Reply by mscofield0** [reply]
Okay, I will post it some time later.

**Reply by mscofield0** [reply]
Tried with the example from [here](https://arthursonzogni.github.io/FTXUI/examples_2component_2input_8cpp-example.html), it's definitely reproducible (the input not being cleared part):

```cpp
std::string first_name;
std::string last_name;
std::string password;

InputOption opt;
opt.on_enter = [&first_name] { first_name.clear(); };
Component input_first_name = Input(&first_name, "first name", opt);
Component input_last_name = Input(&last_name, "last name");
Component input_password = Input(&password, "password");

auto component = Container::Vertical({
  input_first_name,
  input_last_name,
  input_password,
});

auto renderer = Renderer(component, [&] {
	return vbox({
			 text("Hello " + first_name + " " + last_name),
			 separator(),
			 hbox(text(" First name : "), input_first_name->Render()),
			 hbox(text(" Last name  : "), input_last_name->Render()),
			 hbox(text(" Password   : "), input_password->Render()),
		   }) |
	  border;
});

auto screen = ScreenInteractive::TerminalOutput();
screen.Loop(renderer);
```

**Reply by mscofield0** [reply]
Hmm, this one doesn't show the same results as the original one, but it still doesn't work as expected:
![image](https://user-images.githubusercontent.com/69715407/145110371-f8983966-f2a7-4611-8746-5d94534358ca.png)

Code:
```cpp
std::string first_name;
std::string last_name;
std::string password;

Component list = Container::Vertical({});

InputOption opt;
opt.on_enter = [&first_name, &list] {
	list->Add(Renderer([&first_name] { return text(first_name); }));
	first_name.clear();
};
Component input_first_name = Input(&first_name, "first name", opt);
Component input_last_name = Input(&last_name, "last name");
Component input_password = Input(&password, "password");

auto component = Container::Vertical({
  input_first_name,
  input_last_name,
  input_password,
});

auto renderer = Renderer(component, [&] {
	return vbox({
			 window(text("List"), list->Render()),
			 text("Hello " + first_name + " " + last_name),
			 separator(),
			 hbox(text(" First name : "), input_first_name->Render()),
			 hbox(text(" Last name  : "), input_last_name->Render()),
			 hbox(text(" Password   : "), input_password->Render()),
		   }) |
	  border;
});

auto screen = ScreenInteractive::TerminalOutput();
screen.Loop(renderer);
```

Update:
This can be fixed by passing a copy into the `Renderer()` lambda. Don't know why though, `text()` takes in a copy of the string, not a reference...

**Reply by mscofield0** [reply]
Okay fixed my original issue with output being erased. Apparently you can't move a `std::string` into `text()` for some reason.

**Reply by mscofield0** [reply]
Though I would like to find out a solution to erasing the input. @ArthurSonzogni

**Reply by ArthurSonzogni** [reply]
About `Input`, there is known bug. Trying to modify the string value during the on_enter event is ineffective.
https://github.com/ArthurSonzogni/FTXUI/discussions/271

What can be done:
1. Wait for me to fix it.
2. Use the deprecated constructor from `std::wstring` instead of `std::string`.
3. Delay mutating the input into another task, outside of the on_enter function.

The issue is that the Input implementation is still made using std::wstring, and we "synchronize" manually back and forth during the OnEvent event:
https://github.com/ArthurSonzogni/FTXUI/blob/master/src/ftxui/component/input.cpp#L213
